web/siteplan/src/util/Controls/ExportControl.ts

```diff
@@ -50,182 +70,417 @@ export default class ExportControl extends Control {
-    const width = Math.round((ExportControl.PAPER_SIZE[0] * RESOLUTION_DPI) / 25.4)
-    const height = Math.round((ExportControl.PAPER_SIZE[1] * RESOLUTION_DPI) / 25.4)
+    outsidePolygonFeature.setStyle(new Style({
+      stroke: undefined
+    }))
+    return outsidePolygonFeature
+  }
 
-    // Draw layers to canvas
-    const canvas = document.createElement('canvas')
-    canvas.width = ExportControl.PAPER_SIZE[0] * RENDER_RESOLUTION
-    canvas.height = ExportControl.PAPER_SIZE[1] * RENDER_RESOLUTION
-    const context = canvas.getContext('2d')
+  private async createTileCanvas (
+    tilesdata: ExportTileData[],
+    visibleLayers: NamedFeatureLayer[],
+    rotationData?: RotateData
+  ) {
+    const viewsize = this.map.getSize()
+    if (!viewsize) {
+      return
+    }
 
-    if (context == null) {
+    const horizontalRotate: RotateData | undefined = rotationData ? {
+      rad: rotationData.rad - toRad(90),
+      anchor: rotationData.anchor
+    } : undefined
+    const resolution = this.map.getView().getResolution() as number
+    this.map.getView().setRotation(horizontalRotate ? horizontalRotate.rad : 0)
+    const sleep = (ms: number) => new Promise(res => setTimeout(res, ms))
+    const canvas = document.createElement('canvas')
+    const tileTotalSize = this.getTileTotalSize(tilesdata)
+    canvas.width = tileTotalSize.width
+    canvas.height = tileTotalSize.height
+    const context = canvas?.getContext('2d')
+    if (context == null ){
       return
     }
 
     context.fillStyle = 'white'
     context.fillRect(0, 0, canvas.width, canvas.height)
+    for (const tileData of tilesdata) {
+      const [tileMinX,,, tileMaxY] = tileData.tileExtent
+      const rotatedCenter = horizontalRotate
+        ? pointRotate(tileData.center, toDeg(horizontalRotate.rad), horizontalRotate.anchor)
+        : tileData.center
+      const x = (tileMinX - tileTotalSize.minX) / resolution
+      const y = (tileTotalSize.maxY - tileMaxY) / resolution
+      const tileLayer = this.getTileLayer(
+        tileData,
+        visibleLayers,
+        horizontalRotate
+      )
+      if (tileLayer == null) {
+        continue
+      }
 
-    this.map.once('rendercomplete', () => {
-      const canvasLayers = document.querySelectorAll('.ol-layer canvas')
-      canvasLayers.forEach(layerelement => {
-        const layer = layerelement as HTMLCanvasElement
-        if (layer.width <= 0) {
-          return
-        }
-
-        const context = canvas.getContext('2d')
-        if (context === null) {
-          return
-        }
+      this.map.getLayers().push(tileLayer)
+      this.map.getView().setCenter(rotatedCenter)
 
-        context.drawImage(layer, 0, 0)
+      this.map.once('rendercomplete', () => {
+        const canvasLayers = document.querySelectorAll('.ol-layer canvas')
+        canvasLayers.forEach(layer => this.drawTileLayer(layer as HTMLCanvasElement, canvas, x, y, viewsize))
       })
+      await sleep(2000)
```

Comment:
@TruongQuangSB I think you need the sleep to wait for the `rendercomplete` event. This can be solved easier by

```suggestion
      await new Promise((resolve) => {
        this.map.once('rendercomplete', () => {
          const canvasLayers = document.querySelectorAll('.ol-layer canvas')
          canvasLayers.forEach(layer => this.drawTileLayer(layer as HTMLCanvasElement, canvas, x, y, viewsize))
          resolve();
        })
      })
```